_sources/configuration.rst.txt
==================================
Конфигурация
==================================

Конфигурация службы LLDPd_ осуществляется в подразделе "Настройки"
(раздел главного меню :menuselection:`Службы --> LLDP`).

.. image:: images/config-page.png
    :width: 100%
    :class: with-border

Страница настрек разделена на следующие вкладки:

- "Основные настройки" (см. :numref:`sec-config-general`)
- "Сетевые интерфейсы" (см. :numref:`sec-config-network`)
- "Дополнительные настройки" (см. :numref:`sec-config-advanced`)
- "Поддержка протоколов" (см. :numref:`sec-config-protocols`)


.. _sec-config-general:

Основные настройки
==================

Внешний вид вкладки "Основные настройки" показан на :numref:`fig-config-tab-general`.

.. _fig-config-tab-general:
.. figure:: images/config-tab-general.png
    :width: 100%
    :class: with-border

    Вкладка "Основные настройки"

Включить службу
---------------

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - *включено*
   * - UCI-опция
     - *нет*

Включение или отключение автоматического запуска службы LLDPd_ при запуске системы.

Описание системы
-----------------

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - *нет*
   * - UCI-опция
     - ``lldpd.config.lldp_description``

Текстовое описание системы. Данное значение будет передаваться другим устройствам
по сети через включенные протоколы обнаружения.

.. _sec-config-general-sysname:

Имя системы
---------------

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - *нет*
   * - UCI-опция
     - ``lldpd.config.lldp_hostname``

Имя системы. Данное значение будет передаваться другим устройствам
по сети через включенные протоколы обнаружения.

.. _sec-config-general-sysdescr:

Описание платформы
-------------------

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - *нет*
   * - UCI-опция
     - ``lldpd.config.lldp_platform``

Текстовое описание платформы. Данное значение будет передаваться другим устройствам
по сети через включенные протоколы обнаружения.

Если значение не указано, будет использовано имя ядра системы (Linux).

.. _sec-config-general-mgmt-ip:

Адреса управления данной системы
---------------------------------

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - *нет*
   * - UCI-опция
     - ``lldpd.config.lldp_sys_mgmt_ip``

Адреса управления данной системой. Если не указано, то используется первый IPv4
и первый IPv6 адреса. Если указан конкретный IP-адрес, то он будет использован как адрес
управления без какой-либо проверки. Если необходимо занести в черный список
IPv6-адреса, то можно использовать шаблон ``! *: *``.

.. note::
  Подробная информация о доступных шаблонах данного параметра
  приведена в `официальной документации <https://vincentbernat.github.io/lldpd/usage.html>`_
  службы LLDPd_.

Задержка отправки
--------------------

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - ``30``
   * - UCI-опция
     - ``lldpd.config.lldp_tx_interval``

Задержка между двумя передачами LLDP PDU. Значение по умолчанию
составляет 30 секунд.

Значение "transmit hold"
---------------------------

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - ``4``
   * - UCI-опция
     - ``lldpd.config.lldp_tx_hold``

Это значение используется для вычисления :term:`TTL` переданных пакетов,
которое является произведением этого значения и задержки передачи.
Значение по умолчанию равно 4. Соответственно значение :term:`TTL` по
умолчанию составляет 120 секунд.

Включить режим "только приём"
-------------------------------

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - ``0`` (отключено)
   * - UCI-опция
     - ``lldpd.config.readonly_mode``

Управление режимом "только приём". В режиме "только приём"
служба LLDPd_ не выполняет отправку пакетов. Выполняется
только приём и анализ пакетов.


.. _sec-config-network:

Сетевые интерфейсы
==================

Внешний вид вкладки "Сетевые интерфейсы" показан на :numref:`fig-config-tab-network`.

.. _fig-config-tab-network:
.. figure:: images/config-tab-network.png
    :width: 100%
    :class: with-border

    Вкладка "Сетевые интерфейсы"

Сетевые интерфейсы
------------------

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - *нет*
   * - UCI-опция
     - ``lldpd.config.interface``

Выбор сетевых интерфейсов, которые будут прослушивать и
отправлять пакеты :term:`LLDPDU`. Если не указан ни один интерфейс,
служба LLDPd_ будет использовать все доступные физические интерфейсы.

.. _sec-config-network-mgmt-ifaces:

Сетевые интерфейсы, используемые для вычисления ID шасси (chassis ID)
---------------------------------------------------------------------

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - *нет*
   * - UCI-опция
     - ``lldpd.config.cid_interface``

Выбор сетевых интерфейсов, которые должны использоваться для
вычисления идентификатора шасси (chassis ID). Если интерфейсы
не указаны, то будут иcпользованы все интерфейсы. В этом случае
служба LLDPd_ будет использовать первый MAC-адрес из всех доступных
интерфейсов для вычисления ID шасси.


.. _sec-config-advanced:

Дополнительные настройки
========================

Внешний вид вкладки "Дополнительные настройки" показан на :numref:`fig-config-tab-advanced`.

.. _fig-config-tab-advanced:
.. figure:: images/config-tab-advanced.png
    :width: 100%
    :class: with-border

    Вкладка "Дополнительные настройки"

Путь к сокету SNMP agentX
-------------------------

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - ``/var/run/agentx.sock``
   * - UCI-опция
     - ``lldpd.config.agentxsocket``

Если указан путь к сокету :term:`SNMP` агента, то служба
LLDPd_ включит :term:`SNMP` агент, используя протокол AgentX.
Это позволяет получать информацию о локальной системе и
удаленных системах по протоколу :term:`SNMP`.


Класс устройства LLDP-MED
-------------------------

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - ``4``
   * - UCI-опция
     - ``lldpd.config.lldp_class``

Выбор класса устройства LLDP-MED. Доступные значения приведены в таблице:

.. list-table::
   :class: full-width
   :header-rows: 1

   * - Значение LuCI
     - Значение UCI
   * - Generic Endpoint (класс I)
     - ``1``
   * - Media Endpoint (класс II)
     - ``2``
   * - Communication Device Endpoints (класс III)
     - ``3``
   * - Network Connectivity Device (класс IV)
     - ``4``

Отключить отправку инвентарной информации LLDP-MED TLV
------------------------------------------------------

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - ``0`` (отключено)
   * - UCI-опция
     - ``lldpd.config.lldpmed_no_inventory``

Отключение отправки инвентарной информации :term:`LLDP-MED` :term:`TLV`.

Служба LLDPd_ будет по прежнему получать (и публиковать,
используя :term:`SNMP`, если он включен) информацию :term:`LLDP-MED` :term:`TLV`,
но не будет её отправлять. Используйте эту опцию, если вы не
хотите передавать важную информацию, например, такую как серийные номера.

Поведение при обнаружении нескольких соседей
--------------------------------------------

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - ``15``
   * - UCI-опция
     - ``lldpd.config.filter``

Определяет поведение при обнаружении нескольких соседей на одном порту.

В гетерогенных сетях вы можете видеть несколько различных хостов на
одном и том же порту, даже если к этому порту физически подключен только один.
Например, если у вас есть коммутатор Nortel с запущенным :term:`LLDP`, который
подключен к коммутатору Cisco, работающему по протоколу :term:`CDP`, и ваш
хост подключен к Cisco коммутатору, вы увидите также и Nortel коммутатор,
потому что :term:`LLDP` кадры будут перенаправляться коммутатором Cisco.

Входящая и исходящая фильтрация не связаны между собой.
Входящая фильтрация позволяет скрывать некоторые удаленные порты,
чтобы вы могли точно знать, какое оборудование находится по другую
сторону сетевого кабеля. Фильтрация исходящих кадров позволит запретить
использование некоторых протоколов, чтобы избежать флудинга сети протоколами,
которые никак не обрабатываются ближайшим оборудованием.

.. note::
  Следует иметь в виду, что без фильтрации служба LLDPd_ будет
  "общаться" по всем протоколам, для которых был получен хотя бы
  один кадр, а :term:`LLDP` --- нет.

При включении входящей фильтрации служба LLDPd_ попытается выбрать
один протокол и отфильтровать соседей, использующих другие протоколы.
Для выбора протокола правило заключается в том, чтобы взять менее используемый
протокол. Если на одном порту вы получаете 12 соседей :term:`CDP` и 1 соседа :term:`LLDP`,
то это значит, что удаленный коммутатор "общается" на :term:`LLDP` и не фильтрует :term:`CDP`.
Поэтому выбирается :term:`LLDP`. При включении исходящей фильтрации служба LLDPd_
также попытается выбрать один протокол и будет "общаться" только по этому
протоколу. Фильтрация производится по портам. Каждый порт может иметь
различный протокол.

При включении фильтрации имеется два дополнительных критерия:

- разрешение выбрать один или несколько протоколов (в случае ничьей);
- разрешение выбрать одного или нескольких соседей.

Правило выбора протоколов с меньшим количеством соседей применяется
Даже если разрешено несколько протоколов. Если служба LLDPd_
выбирает :term:`LLDP` и :term:`CDP`, это означает, что у них одинаковое количество соседей.
Выбор соседей происходит случайным образом. Входящая фильтрация будет
выбирать набор соседей, которые будут отображаться. Исходящая фильтрация
будет использовать выбранный набор соседей, чтобы решить какие протоколы
использовать: если выбранный сосед "общается" через :term:`LLDP`,
а другой --- через :term:`CDP`, то LLDPd_ будет "общаться"
и через :term:`CDP`, и через :term:`LLDP` на этом порту.

Есть несколько граничных случаев. Типичный пример --- коммутатор, "общающийся"
на двух протоколах (например, :term:`CDP` и :term:`LLDP`). Вы хотите получить информацию
по лучшему протоколу, но хотите "общаться" на обоих протоколах, потому что
некоторые инструменты используют таблицу :term:`CDP`, а другие --- :term:`LLDP`.

В таблице ниже приведены все возможные значения фильтрации.
Значение по умолчанию равно 15, что соответствует граничному случаю,
описанному выше. Столбец "filter" означает, что фильтрация включена.
Столбец "1proto" говорит о том, что будет сохранен только один протокол.
Столбец "1neigh" говорит о том, что будет сохранен только один сосед.

.. table:: Возможные значения фильтрации входящих и исходящих пакетов
    :class: full-width

    +------------+--------------------------+--------------------------+
    |            | Входящая фильтрация      | Исходящая фильтрация     |
    | Значение   +--------+--------+--------+--------+--------+--------+
    |            | filter | 1proto | 1neigh | filter | 1proto | 1neigh |
    +============+========+========+========+========+========+========+
    | 0          |        |        |        |        |        |        |
    +------------+--------+--------+--------+--------+--------+--------+
    | 1          | ✔      | ✔      |        | ✔      | ✔      |        |
    +------------+--------+--------+--------+--------+--------+--------+
    | 2          | ✔      | ✔      |        |        |        |        |
    +------------+--------+--------+--------+--------+--------+--------+
    | 3          |        |        |        | ✔      | ✔      |        |
    +------------+--------+--------+--------+--------+--------+--------+
    | 4          | ✔      |        |        | ✔      |        |        |
    +------------+--------+--------+--------+--------+--------+--------+
    | 5          | ✔      |        |        |        |        |        |
    +------------+--------+--------+--------+--------+--------+--------+
    | 6          |        |        |        | ✔      |        |        |
    +------------+--------+--------+--------+--------+--------+--------+
    | 7          | ✔      | ✔      | ✔      | ✔      | ✔      |        |
    +------------+--------+--------+--------+--------+--------+--------+
    | 8          | ✔      | ✔      | ✔      |        |        |        |
    +------------+--------+--------+--------+--------+--------+--------+
    | 9          | ✔      |        | ✔      | ✔      | ✔      |        |
    +------------+--------+--------+--------+--------+--------+--------+
    | 10         |        |        |        | ✔      |        | ✔      |
    +------------+--------+--------+--------+--------+--------+--------+
    | 11         | ✔      |        | ✔      |        |        |        |
    +------------+--------+--------+--------+--------+--------+--------+
    | 12         | ✔      |        | ✔      | ✔      |        | ✔      |
    +------------+--------+--------+--------+--------+--------+--------+
    | 13         | ✔      |        | ✔      | ✔      |        |        |
    +------------+--------+--------+--------+--------+--------+--------+
    | 14         | ✔      | ✔      |        | ✔      |        | ✔      |
    +------------+--------+--------+--------+--------+--------+--------+
    | 15         | ✔      | ✔      |        | ✔      |        |        |
    +------------+--------+--------+--------+--------+--------+--------+
    | 16         | ✔      | ✔      | ✔      | ✔      |        | ✔      |
    +------------+--------+--------+--------+--------+--------+--------+
    | 17         | ✔      | ✔      | ✔      | ✔      |        |        |
    +------------+--------+--------+--------+--------+--------+--------+
    | 18         | ✔      |        |        | ✔      |        | ✔      |
    +------------+--------+--------+--------+--------+--------+--------+
    | 19         | ✔      |        |        | ✔      | ✔      |        |
    +------------+--------+--------+--------+--------+--------+--------+

.. _sec-config-advanced-portidsubtype:

Использовать в качестве ID порта
--------------------------------

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - ``macaddress``
   * - UCI-опция
     - ``lldpd.config.lldp_portidsubtype``

Позволяет выбрать что использовать в качестве идентификатора
порта --- MAC-адрес или имя интерфейса. Доступные значения приведены в таблице:

===================== ============== ========================================================================
Значение LuCI         Значение UCI   Описание
===================== ============== ========================================================================
MAC-адрес интерфейса  ``macaddress`` В качестве идентификатора порта будет использован MAC-адрес интерфейса.
Имя интерфейса        ``ifname``     В качестве идентификатора порта будет использовано имя интерфейса.
===================== ============== ========================================================================

MAC-адрес назначения, используемый для отправки LLDPDU
------------------------------------------------------

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - ``nearest-bridge``
   * - UCI-опция
     - ``lldpd.config.lldp_agenttype``

MAC-адрес назначения, используемый для отправки :term:`LLDPDU`. Данный параметр позволяет
агенту контролировать распространение :term:`LLDPDU`. По умолчанию используется MAC-адрес
``01:80:c2:00:00:0e`` и ограничивает распространение :term:`LLDPDU` до ближайшего моста.
Доступные значения приведены в таблице:

============================================ =========================== ========================================================================
Значение LuCI                                Значение UCI                Описание
============================================ =========================== ========================================================================
01:80:c2:00:00:0e (nearest-bridge)           ``nearest-bridge``          :term:`LLDPDU` будут отпавляться на MAC-адрес ``01:80:c2:00:00:0e``.
01:80:c2:00:00:03 (nearest-nontpmr-bridge)   ``nearest-nontpmr-bridge``  :term:`LLDPDU` будут отпавляться на MAC-адрес ``01:80:c2:00:00:03``.
01:80:c2:00:00:00 (nearest-customer-bridge)  ``nearest-customer-bridge`` :term:`LLDPDU` будут отпавляться на MAC-адрес ``01:80:c2:00:00:00``.
============================================ =========================== ========================================================================

.. _sec-config-protocols:

Поддержка протоколов
====================

Вкладка "Поддержка протоколов" дополнительно разделена на подвкладки
для настроек каждого из поддерживаемых протоколов:

- :term:`LLDP` (см. :numref:`sec-config-protocols-lldp`);
- :term:`CDP` (см. :numref:`sec-config-protocols-cdp`);
- :term:`FDP` (см. :numref:`sec-config-protocols-fdp`);
- :term:`EDP` (см. :numref:`sec-config-protocols-edp`);
- :term:`SONMP`, :term:`NTDP`, :term:`BNMP`, :term:`NDP`, :term:`BDP` (см. :numref:`sec-config-protocols-sonmp`).

.. _sec-config-protocols-lldp:

Протокол LLDP
-------------

Внешний вид вкладки с настройками протокола :term:`LLDP` показан
на :numref:`fig-config-tab-protocols-lldp`.

.. _fig-config-tab-protocols-lldp:
.. figure:: images/config-tab-protocols-lldp.png
    :width: 100%
    :class: with-border

    Вкладка "Поддержка протоколов". Настройки протокола :term:`LLDP`

Включить LLDP
^^^^^^^^^^^^^

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - ``1`` (включено)
   * - UCI-опция
     - ``lldpd.config.enable_lldp``

Опция включает поддержку протокола :term:`LLDP`.

Форсировать отправку LLDP пакетов
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

.. list-table::
   :widths: 30 70
   :class: full-width
   :header-rows: 0

   * - Значение по умолчанию
     - ``0`` (отключено)
   * - UCI-опция
     - ``lldpd.config.force_lldp``

Опция управляет принудительной отправкой пакетов :term:`LLDP`, даже если соседей
с поддержкой :term:`LLDP` не обнаружено, но есть соседи, с поддержкой
другого протокола. По умолчанию пакеты :term:`LLDP` отправляются,
когда обнаружен сосед с поддержкой :term:`LLDP` или когда вообще
нет обнаруженных соседей.

.. _sec-config-protocols-cdp:

Протокол CDP
------------

Внешний вид вкладки с настройками протокола :term:`CDP` показан
на :numref:`fig-config-tab-protocols-cdp`.

.. _fig-config-tab-protocols-cdp:

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

All dates in this document are in `DD.MM.YYYY` format.

## [Unreleased]

**Nothing yet**

## [Version 2.0.1] (06.12.2020)
### Fixed
- Fixed IE11 compatibility.
- Fixed Rx interface statistics display.

## [Version 2.0.0] (22.04.2020)
### Changed
- Coverted to client side rendering.

## [Version 1.3.0] (21.04.2020)
### Added
- Added ACL rules for latest LuCI support
- Allow to hide footer by UCI option 'luci.app_tn_lldpd.hide_footer'

### Changed
- Changed application pages titles.
- All tables wrapped into div with class "table-wrapper".
- Use polling interval from LuCI configuration (luci.main.pollinterval)
- Added spinner for messages about waiting for data.
- Removed useless ucitrack definition
- Use L.Poll.add() instead of deprecated XHR.Poll() for data polling.

### Fixed
- Fixed possible simultaneous folding/unfolding of multiple rows.
- Dropped unnecessary data from the "discovered port" column
  of the neighbors table.
- Minor HTML markup fixes.
- Handle JSON.parse() exceptions.

## [Version 1.2.1] (25.12.2018)
### Added
- Added support for the official OpenWrt and Material LuCI themes.
  Previously supported only Bootstrap LuCI theme.
- Enable/disable service checkbox.
- Additional JavaScript checks for case when LLDPd not started.

### Changed
- Update screenshots for neighbors and statistics tables.
- Badges with individual colors for all supported protocols in neighbors
  and statistics tables. Previously inidividual badges will available
  only for LLDP and CDP protocols. EDP, FDP and SONMP protocols had the
  same colors for badges.
- Changed icons for show/hide additional information.
- Fold and unfold table cells when clicking on any place in the row.

## [Version 1.2.0] (15.12.2018)
### Added
- This CHANGELOG.md file.
- New screenshots.
- Extra validators for parameters.
- Local chassis information view.
- The ability to configure the following new parameters:
  * Port ID subtype applied to all interfaces (MAC or name of the interface)
  * The destination MAC address used to send LLDPDU
  * Receive-only mode (lldpd -r parameter)
  * Interfaces to use for computing chassis ID (lldpd -C parameter)
  * Filter neighbors (lldpd -H parameter)
  * Disable LLDP-MED inventory TLV transmission (lldpd -i parameter)
- Extra comments for existing parameters.

### Changed
- Improved tables appearance and usability. Allow to collapse/expand additional information
  for each table row.
- Merged status and config menu items as 3rd level pages under "Services"
  main menu item.

### Fixed
- HTML markup and JavaScript fixes.

## [Version 1.1.1] (18.09.2018)
### Fixed
- Fixed lldpd service reloading.
- Some HTML markup fixes.

## [Version 1.1.0] (17.07.2018)
### Added
- Added support for LuCI from OpenWrt 18.06.x branch.

## [Version 1.0.0] (05.07.2018)

Initial release

[Unreleased]: https://github.com/tano-systems/luci-app-tn-lldpd/tree/master
[Version 2.0.1]: https://github.com/tano-systems/luci-app-tn-lldpd/releases/tag/v2.0.1
[Version 2.0.0]: https://github.com/tano-systems/luci-app-tn-lldpd/releases/tag/v2.0.0
[Version 1.3.0]: https://github.com/tano-systems/luci-app-tn-lldpd/releases/tag/v1.3.0
[Version 1.2.2]: https://github.com/tano-systems/luci-app-tn-lldpd/releases/tag/v1.2.2
[Version 1.2.1]: https://github.com/tano-systems/luci-app-tn-lldpd/releases/tag/v1.2.1
[Version 1.2.0]: https://github.com/tano-systems/luci-app-tn-lldpd/releases/tag/v1.2.0
[Version 1.1.1]: https://github.com/tano-systems/luci-app-tn-lldpd/releases/tag/v1.1.1
[Version 1.1.0]: https://github.com/tano-systems/luci-app-tn-lldpd/releases/tag/v1.1.0
[Version 1.0.0]: https://github.com/tano-systems/luci-app-tn-lldpd/releases/tag/v1.0.0


luasrc/model/cbi/lldpd/config.lua
--
-- Copyright (c) 2018, Tano Systems. All Rights Reserved.
-- Anton Kikin <a.kikin@tano-systems.com>
--

local sys  = require "luci.sys"
local util = require "luci.util"

local m, s, o

m = Map("lldpd",
	translate("LLDPd Settings"),
	translate(
		"LLDPd is a implementation of IEEE 802.1ab " ..
		"(<abbr title=\"Link Layer Discovery Protocol\">LLDP</abbr>)."
	) ..
	translate(
		"On this page you may configure LLDPd parameters."
	)
)

s = m:section(TypedSection, "lldpd")
s.anonymous = true
s.addremove = false

s:tab("basic", translate("Basic Settings"))
s:tab("ifaces", translate("Network Interfaces"))
s:tab("advanced", translate("Advanced Settings"))
s:tab("protocols", translate("Protocols Support"))

-----------------------------------------------------------------------------------
--
-- Basic settings
--
-----------------------------------------------------------------------------------

-- Service enable/disable
o = s:taboption("basic", Flag, "enabled",
	translate("Enable service")
)

o.rmempty = false

function o.cfgvalue(self, section)
	return sys.init.enabled("lldpd") and self.enabled or self.disabled
end

function o.write(self, section, value)
	if value == "1" then
		sys.init.enable("lldpd")
		sys.call("/etc/init.d/lldpd start >/dev/null 2>&1")
	else
		sys.call("/etc/init.d/lldpd stop >/dev/null 2>&1")
		sys.init.disable("lldpd")
	end

	return Flag.write(self, section, value)
end

-- System description
o = s:taboption("basic", Value, "lldp_description",
	translate("System description"),
	translate(
		"Override system description with the provided description."
	)
)

o.placeholder = "System description"

-- System hostname
o = s:taboption("basic", Value, "lldp_hostname",
	translate("System hostname"),
	translate(
		"Override system hostname with the provided value."
	)
)

o.placeholder = "System hostname"

-- Platform
o = s:taboption("basic", Value, "lldp_platform",
	translate("System platform description"),
	translate(
		"Override the platform description with the provided value. " ..
		"The default description is the kernel name (Linux)."
	)
)

o.placeholder = "System platform description"

-- Management addresses of this system
o = s:taboption("basic", Value, "lldp_sys_mgmt_ip",
	translate("Management addresses of this system"),
	translate(
		"Specify the management addresses of this system. " ..
		"If not specified, the first IPv4 and the first " ..
		"IPv6 are used. If an exact IP address is provided, it is used " ..
		"as a management address without any check. If you want to " ..
		"blacklist IPv6 addresses, you can use <code>!*:*</code>. " ..
		"See more details about available patterns " ..
		"<a href=\"https://vincentbernat.github.io/lldpd/usage.html\">here</a>."
	)
)

o.placeholder = "Management addresses"

-- LLDP tx interval
o = s:taboption("basic", Value, "lldp_tx_interval",
	translate("Transmit delay"),
	translate(
		"The transmit delay is the delay between two " ..
		"transmissions of LLDP PDU. The default value " ..
		"is 30 seconds."
	)
)

o.datatype = "uinteger"
o.default = 30
o.placeholder = 30
o.rmempty = false

function o.validate(self, value, section)
	if tonumber(value) <= 0 then
		return nil, translate("Transmit delay must be greater than 0")
	end
	return Value.validate(self, value, section)
end

-- LLDP tx hold
o = s:taboption("basic", Value, "lldp_tx_hold",
	translate("Transmit hold value"),
	translate(
		"This value is used to compute the TTL of transmitted " ..
		"packets which is the product of this value and of the " ..
		"transmit delay. The default value is 4 and therefore " ..
		"the default TTL is 120 seconds."
	)
)

o.datatype = "uinteger"
o.default = 4
o.placeholder = 4
o.rmempty = false

function o.validate(self, value, section)
	if tonumber(value) <= 0 then 
		return nil, translate("Transmit hold value must be greater than 0")
	end
	return Value.validate(self, value, section)
end

-- Received-only mode (-r)
o = s:taboption("basic", Flag, "readonly_mode",
	translate("Enable receive-only mode"),
	translate(
		"With this option, LLDPd will not send any frames. " ..
		"It will only listen to neighbors."
	)
)

o.default = "0"

-----------------------------------------------------------------------------------
--
-- Network Interfaces
--
-----------------------------------------------------------------------------------

-- Interfaces to listen on
o = s:taboption("ifaces", DynamicList, "interface",
	translate("Network interfaces"),
	translate(
		"Specify which interface to listen and send LLDPDU to." ..
		"If no interfaces is specified, LLDPd will use all available physical interfaces."
	)
)

o.template  = "cbi/network_ifacelist"
o.nobridges = true
o.rmempty   = true
o.nocreate  = true
o.network   = ""
o.widget    = "checkbox"

-- ChassisID interfaces
o = s:taboption("ifaces", DynamicList, "cid_interface",
	translate("Network interfaces for chassis ID computing"),
	translate(
		"Specify which interfaces to use for computing chassis ID. " ..
		"If no interfaces is specified, all interfaces are considered. " ..
		"LLDPd will take the first MAC address from all the considered " ..
		"interfaces to compute the chassis ID."
	)
)

o.template  = "cbi/network_ifacelist"
o.nobridges = false
o.rmempty   = true
o.nocreate  = true
o.network   = ""
o.widget    = "checkbox"


-----------------------------------------------------------------------------------
--
-- Advanced Settings
--
-----------------------------------------------------------------------------------

-- SNMP agentX socket
o = s:taboption("advanced", Value, "agentxsocket",
	translate("SNMP agentX socket path"),
	translate(
		"If the path to the socket is set, then LLDPd will enable an " ..
		"SNMP subagent using AgentX protocol. This allows you to get " ..
		"information about local system and remote systems through SNMP."
	)
)

o.rmempty = false
o.placeholder = "/var/run/agentx.sock"
o.default = "/var/run/agentx.sock"


-- LLDP class
o = s:taboption("advanced", ListValue, "lldp_class",
	translate("LLDP-MED device class")
)

o:value(1, translate("Generic Endpoint (Class I)"))
o:value(2, translate("Media Endpoint (Class II)"))
o:value(3, translate("Communication Device Endpoints (Class III)"))
o:value(4, translate("Network Connectivity Device (Class IV)"))

o.default = 4

-- LLDP-MED inventory TLV transmission (-i)
o = s:taboption("advanced", Flag, "lldpmed_no_inventory",
	translate("Disable LLDP-MED inventory TLV transmission"),
	translate(
		"LLDPd will still receive (and publish using SNMP if enabled) " ..
		"those LLDP-MED TLV but will not send them. Use this option " ..
		"if you don't want to transmit sensible information like serial numbers."
	)
)

o.default = "0"

-- Filter neighbors (-H)
o = s:taboption("advanced", Value, "filter",
	translate("Specify the behaviour when detecting multiple neighbors"),
	translate(
		"The default filter is 15. For more details see \"FILTERING NEIGHBORS\" section " ..
		"<a href=\"https://vincentbernat.github.io/lldpd/usage.html\">here</a>."
	)
)

o.template = "lldpd/filter_select"
o.default  = 15

-- Force port ID subtype
o = s:taboption("advanced", ListValue, "lldp_portidsubtype",
	translate("Force port ID subtype"),
	translate(
		"With this option, you can force the port identifier " ..
		"to be the interface name or the MAC address."
	)
)

o:value("macaddress", translate("Interface MAC address"))
o:value("ifname", translate("Interface name"))

o.default = "macaddress"

-- The destination MAC address used to send LLDPDU
o = s:taboption("advanced", ListValue, "lldp_agenttype",
	translate("The destination MAC address used to send LLDPDU"),
	translate(
		"The destination MAC address used to send LLDPDU allows an agent " ..
		"to control the propagation of LLDPDUs. By default, the " ..
		"<code>01:80:c2:00:00:0e</code> MAC address is used and limit the propagation " ..
		"of the LLDPDU to the nearest bridge."
	)
)

o:value("nearest-bridge",          "01:80:c2:00:00:0e (nearest-bridge)")
o:value("nearest-nontpmr-bridge",  "01:80:c2:00:00:03 (nearest-nontpmr-bridge)")
o:value("nearest-customer-bridge", "01:80:c2:00:00:00 (nearest-customer-bridge)")

o.default = "nearest-bridge"

-----------------------------------------------------------------------------------
--
-- Protocols Support
--
-----------------------------------------------------------------------------------

local function make_proto_subtitle(s, title)
	local t = s:taboption("protocols", DummyValue, "_cdp", "&#160;")
	t.default = '<h3>' .. title .. "</h3>"
	t.rawhtml = true
end

-- LLDP
make_proto_subtitle(s, translate("LLDP protocol"))

o = s:taboption("protocols", Flag, "enable_lldp",
	translate("Enable LLDP")
)

o.default = "1"
o.rmempty = true

--
o = s:taboption("protocols", Flag, "force_lldp",
	translate("Force to send LLDP packets"),
	translate(
		"Force to send LLDP packets even when there is no LLDP peer " ..
		"detected but there is a peer speaking another protocol detected. " ..
		"By default, LLDP packets are sent when there is a peer speaking " ..
		"LLDP detected or when there is no peer at all."
	)
)

o.default = "0"
o.rmempty = true
o:depends("enable_lldp", true)

-- CDP
make_proto_subtitle(s, translate("CDP protocol"))

o = s:taboption("protocols", Flag, "enable_cdp",
	translate("Enable CDP"),
	translate(
		"Enable the support of CDP protocol to deal with Cisco routers " ..
		"that do not speak LLDP"
	)
)

o.default = "1"
o.rmempty = false

--
o = s:taboption("protocols", ListValue, "cdp_version",
	translate("CDP version"))

o:value("cdpv1v2", translate("CDPv1 and CDPv2"))
o:value("cdpv2",   translate("Only CDPv2"))
o:depends("enable_cdp", true)

o.default = "cdpv1v2"

--
o = s:taboption("protocols", Flag, "force_cdp",
	translate("Send CDP packets even if no CDP peer detected"))

o.default = "0"
o.rmempty = true
o:depends("enable_cdp", true)

o = s:taboption("protocols", Flag, "force_cdpv2",
	translate("Force sending CDPv2 packets"))

o.default = "0"
o.rmempty = true
o:depends({
	force_cdp = true,
	enable_cdp = true,
	cdp_version = "cdpv1v2"
})

-- FDP
make_proto_subtitle(s, translate("FDP protocol"))

o = s:taboption("protocols", Flag, "enable_fdp",
	translate("Enable FDP"),
	translate(
		"Enable the support of FDP protocol to deal with Foundry routers " ..
		"that do not speak LLDP"
	)
)

o.default = "1"
o.rmempty = false

--
o = s:taboption("protocols", Flag, "force_fdp",
	translate("Send FDP packets even if no FDP peer detected"))

o.default = "0"
o.rmempty = true
o:depends("enable_fdp", true)

-- EDP
make_proto_subtitle(s, translate("EDP protocol"))

o = s:taboption("protocols", Flag, "enable_edp",
	translate("Enable EDP"),
	translate(
		"Enable the support of EDP protocol to deal with Extreme routers " ..
		"and switches that do not speak LLDP."
	)
)

o.default = "1"
o.rmempty = false

--
o = s:taboption("protocols", Flag, "force_edp",
	translate("Send EDP packets even if no EDP peer detected"))

o.default = "0"
o.rmempty = true
o:depends("enable_edp", true)

-- SONMP
make_proto_subtitle(s, translate("SONMP protocol"))

o = s:taboption("protocols", Flag, "enable_sonmp",
	translate("Enable SONMP"),
	translate(
		"Enable the support of SONMP protocol to deal with Nortel " ..
		"routers and switches that do not speak LLDP."
	)
)

o.default = "1"
o.rmempty = false

--
o = s:taboption("protocols", Flag, "force_sonmp",
	translate("Send SONMP packets even if no SONMP peer detected"))

o.default = "0"
o.rmempty = true
o:depends("enable_sonmp", true)

-----------------------------------------------------------------------------------

local hide_footer = m.uci:get_bool("luci", "app_tn_lldpd", "hide_footer") or false
if hide_footer == false then
	s = m:section(SimpleSection, nil)
	s.template = "lldpd/footer"
end

return m
